test/internal/test_emerge_planner.py

```diff
@@ -0,0 +1,578 @@
+        source_copy = CopySource(self.recommended_size)
+        write_intents = WriteIntent.wrap_sources_iterator(
+            [source_upload, source_small_copy1, source_small_copy2, source_copy]
+        )
+
+        self.verify_emerge_plan_for_write_intents(
+            write_intents,
+            [
+                part(source_upload),
+                part([
+                    source_small_copy1,
+                    source_small_copy2,
+                ]),
+                part(source_copy),
+            ],
+        )
+
+    def test_upload_multiple_sources(self):
+        self.assertEqual(self.recommended_size % 8, 0)
+
+        unit_part_size = int(self.recommended_size / 8)
+        uneven_part_size = 3 * unit_part_size
+        sources = [
+            UploadSource(uneven_part_size)
+            for i in range(8)
+        ]
+
+        self.verify_emerge_plan_for_write_intents(
+            WriteIntent.wrap_sources_iterator(sources),
+            [
+                part([
+                    sources[0],
+                    sources[1],
+                    (sources[2], 0, 2 * unit_part_size),
+                ]),
+                part([
+                    (sources[2], 2 * unit_part_size, unit_part_size),
+                    sources[3],
+                    sources[4],
+                    (sources[5], 0, unit_part_size),
+                ]),
+                part([
+                    (sources[5], unit_part_size, 2 * unit_part_size),
+                    sources[6],
+                    sources[7],
+                ]),
+            ],
+        )
+
+    def test_small_upload_not_enough_copy_then_upload(self):
+        self.assertGreater(self.min_size, 2 * MEGABYTE)
+
+        source_small_upload = UploadSource(self.min_size - 2 * MEGABYTE)
+        source_copy = CopySource(self.min_size + MEGABYTE)
+        source_upload = UploadSource(self.recommended_size)
+
+        write_intents = WriteIntent.wrap_sources_iterator(
+            [source_small_upload, source_copy, source_upload]
+        )
+        small_parts_len = source_small_upload.get_content_length() + source_copy.get_content_length()
+        source_upload_split_offset = self.recommended_size - small_parts_len
+
+        self.verify_emerge_plan_for_write_intents(
+            write_intents,
+            [
+                part([
+                    source_small_upload,
+                    source_copy,
+                    (source_upload, 0, source_upload_split_offset),
+                ]),
+                part(source_upload, source_upload_split_offset, small_parts_len),
+            ],
+        )
+
+    def test_basic_local_overlap(self):
+        source1 = UploadSource(self.recommended_size * 2)
+        source2 = UploadSource(self.recommended_size * 2)
+        write_intents = [
+            WriteIntent(source1),
+            WriteIntent(source2, destination_offset=self.recommended_size),
+        ]
+
+        self.verify_emerge_plan_for_write_intents(
```

Comment:
I think there should be two uploads but there are many ways in which they can be divided and the test shouldn't fail for a slightly different implementation of planner. In other words this test is too strict. A few tests below are affected by this too.

Reply:
we agreed that we can leave those test in such strict state - I personallt feel that on that level of testing it is rather better to have such strict testing to "freeze" current planning logic a bit - to avoid unintentionall changes in the logic